pkg/controllernamespace/namespace/namespace_controller.go

```diff
@@ -0,0 +1,132 @@
+	return &ReconcileNamespace{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("customeradminreconciler-namespace-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// TODO: figure out how to filter on namespaces that are !prefixed by a certain string
+	err = c.Watch(&source.Kind{Type: &corev1.Namespace{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileNamespace{}
+
+// ReconcileNamespace reconciles a CustomerAdminReconciler object
+type ReconcileNamespace struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile receives a namespace event and reconciles rolebindings for
+// customer roles
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	if !common.RestrictedNamespace(request.Name) {
+		// move to appropriate place
+		// If we are namespace we need to reconcile 3 rolebindings:
+		// - customer-admin
+		// - customer-project-admin
+		// - customer-view
+		logrus.Printf("Reconciling Namespace %s/%s", request.Namespace, request.Name)
+
+		// check for existence of namespace
+		ns := &corev1.Namespace{}
+		err := r.client.Get(context.TODO(), request.NamespacedName, ns)
+		if err != nil {
+			// namespace is not found
+			if kerrors.IsNotFound(err) {
+				logrus.Printf("error: %v", err)
+				return reconcile.Result{}, nil
+			}
+		}
+		// do not reconcile. namespace is being deleted
+		if ns.Status.Phase == corev1.NamespaceTerminating {
+			logrus.Printf("namespace is being deleted %v", request.Namespace)
+			return reconcile.Result{}, nil
+		}
+
+		nn := types.NamespacedName{
+			Namespace: request.Name,
+		}
+		for _, rbname := range common.RoleBindingNames {
+			nrb := common.CustomerRoleBindingMap()(rbname)
+			nrb.Namespace = request.Name
+			nn.Name = rbname
+			rb := &rbacv1.RoleBinding{}
+			err := r.client.Get(context.TODO(), nn, rb)
+			if err != nil {
+				if kerrors.IsNotFound(err) {
+					// need to create rolebinding
+					logrus.Printf("Creating rolebinding: %#v", nrb)
+					err = r.client.Create(context.TODO(), &nrb)
+					if err != nil {
+						logrus.Printf("Error Creating: %v", err)
+						return reconcile.Result{}, err
+					}
+					continue
+				}
+
+				return reconcile.Result{}, err
+			}
+			// check that our rolebinding is correct by seeing if the roleref
```

Comment:
I feel like I need to read up, but I suspect that the way this needs to work is:
1) the namespace controller is responsible solely for creating rolebindings when they don't exist in a namespace.  It can create a rolebinding correctly to save the rolebinding controller some effort, but it only really cares that the rolebinding exists to bootstrap the process.
2) the rolebinding controller is responsible for catching changes to rolebindings it watches and reverting them if it doesn't like them.  Also it can try to catch deletions and if it sees one recreate the object.  If it misses a deletion then (1) will deal with it on a relist.

Given the above:
* it is inevitable that there are situations in which (1) and (2) could both race to create a rolebinding at the same time.  So they need to be tolerant of failing to create because they lost the race.
* there is no need for the namespace controller to worry about changes to rolebindings, those will always be picked up by the rolebinding controller.

Thus, all the code below is not necessary, and the above can be simplified to get rid of some of the tabs.